Your DBMS is not a good random number generator

sexwarrior opened the pull request:
This commit refactors `generateValidationCode()` to rely solely on PHP's preferred source of entropy.

The current approach is not only inefficient, but comes with virtually no guarantee of producing an unpredictable code. Database management systems are good at, well, managing databases. They're not really intended as part of your pseudo-random number generator.

Signed-off by: Miłosz Gaczkowski <milosz@omgomg.eu>

Commit:
Your DBMS is not a good random number generator

This commit refactors `generateValidationCode()` to rely solely on PHP's preferred source of entropy.

The current approach is not only inefficient, but comes with virtually no guarantee of producing an unpredictable code. Database management systems are good at, well, managing databases. They're not really intended as part of your pseudo-random number generator.

Signed-off by: Miłosz Gaczkowski <milosz@omgomg.eu>

albertlast commented:
You could than also remove this lines:
https://github.com/SimpleMachines/SMF2.1/blob/2139a62abcac785300fbbcd1768a64fd413f347b/Sources/Subs-Db-postgresql.php#L333-L335

Commit:
No need to request a random number from the DBMS

We will never have a reason to fetch pseudo-random numbers from PostgreSQL

Signed-off by: Miłosz Gaczkowski <milosz@omgomg.eu>

sexwarrior commented:
> You could than also remove this lines

Agreed.

Oldiesmann commented:
random_bytes is only available in PHP 7 or higher, while SMF currently still supports PHP 5.x.

sexwarrior commented:
> 
> 
> random_bytes is only available in PHP 7 or higher, while SMF currently still supports PHP 5.x.

Fair enough, I thought that was just 2.0. My bad.

It looks like SMF 2.1 is already falling back on a [userland implementation](https://github.com/paragonie/random_compat) of `random_bytes()` for PHP5. A copy is [already present in Sources](https://github.com/SimpleMachines/SMF2.1/tree/ed54e36517065b22151783a6b246347efa14e286/Sources/random_compat) and [a helper function exists to handle the fallback](https://github.com/SimpleMachines/SMF2.1/blob/release-2.1/Sources/Load.php#L260-L271).

It seems like the correct solution would therefore be to just use `$smcFunc['random_bytes']()`, analogous to things like [our current password salt generation](https://github.com/SimpleMachines/SMF2.1/blob/dcbbf523fa90d0b90f3ddd5a2e464f23f3111e4e/Sources/LogInOut.php#L120).

Does that sound about right?

Commit:
Use $smcFunc helper function to allow fallback

`random_bytes()` is not supported in older PHP versions. SMF already provides a helper function with a fallback, so let's use it.

Signed-off by: Miłosz Gaczkowski <milosz@omgomg.eu>

Sesquipedalian reviewed (approved):
> It seems like the correct solution would therefore be to just use `$smcFunc['random_bytes']()`, analogous to things like [our current password salt generation](https://github.com/SimpleMachines/SMF2.1/blob/dcbbf523fa90d0b90f3ddd5a2e464f23f3111e4e/Sources/LogInOut.php#L120).
> 
> Does that sound about right?

Yes, it does. 👍
